File: 3075-maximize-happiness-of-selected-children/3075-maximize-happiness-of-selected-children.java
class Solution {
    public long maximumHappinessSum(int[] happiness, int k) {
        long result = 0;
        int count = 0;

        PriorityQueue<Integer> pq = new PriorityQueue<>(Collections.reverseOrder());

        for (int hap : happiness) {
            pq.offer(hap);
        }

        for (int i = 0; i < k; i++) {
            int hap = pq.poll();
            result += Math.max(hap - count, 0);
            count++;
        }

        return result;
    }
}


// class Solution {
//     public long maximumHappinessSum(int[] happiness, int k) {
//         Arrays.sort(happiness);
//         long sum = 0;
//         int n = happiness.length;
//         for (int i = n - 1, j = 0; i >=  n - k; i--, j++) {
//             if (happiness[i] - j > 0) {
//                 sum += happiness[i] - j;
//             } else {
//                 return sum;
//             }
//         }
//         return sum;
//     }
// }

File: LeetCode/3075-maximize-happiness-of-selected-children/3075-maximize-happiness-of-selected-children.cpp
class Solution {
public:
    long long maximumHappinessSum(vector<int>& happiness, int k) {
        priority_queue<int> pq;
        for (int x : happiness) {
            pq.push(x);
        }
        long long sum = 0;
        int i = 0;
        while (k--) {
            if (pq.top() - i <= 0) {
                return sum;
            }
            sum += pq.top() - i;
            pq.pop();
            i++;
        }
        return sum;
    }
};


// class Solution {
// public:
//     long long maximumHappinessSum(vector<int>& happiness, int k) {
//         sort(happiness.begin(), happiness.end(), greater<int>());
//         long long sum = 0;
//         for (int i = 0; i < k; i++) {
//             if (happiness[i] - i > 0) {
//                 sum += happiness[i] - i;
//             } else {
//                 return sum;
//             }
//         }
//         return sum;
//     }
// };